### ISSUE TITLE
[BUG] containerd registry file is empty after upgrading the embedded Rancher to 2.7.5

### ISSUE THREAD

**Issue by bk201** [harvester/harvester#4156]
**Describe the bug**

The `/etc/rancher/rke2/registries.yaml` file become empty after upgrading the embedded Rancher to 2.7.5

**To Reproduce**
1. Create a 1.1.2 cluster.
2. Configure some registries with the [`containerd-registry` setting](https://docs.harvesterhci.io/dev/advanced/settings#containerd-registry). 
3. Upgrade to v1.2.0-rc2.
4. The file `/etc/rancher/rke2/registries.yaml` becomes empty.

**Expected behavior**

`/etc/rancher/rke2/registries.yaml` should preserve.

**Support bundle**

**Environment**
 - Harvester ISO version: 
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630):

**Additional context**

Previously, we managed the registry file with [a SUC Plan](https://github.com/harvester/harvester/blob/master/deploy/charts/harvester/templates/cattle-system.yaml#L27-L80) (`cattle-system/sync-containerd-registry`)
In Rancher 2.7.5, the registry file is managed by Rancher (see the [comment](https://github.com/harvester/harvester/issues/4095#issuecomment-1598989144) about Rancher plan change). Here is feedback from the Rancher developer:
> We would expect you to define the registries through your [clusters.provisioning.cattle.io](http://clusters.provisioning.cattle.io/) object and let Rancher designate them -- otherwise, you will need to override the registry with machineGlobalConfig (if possible, it may not actually be possible but I need to look at what we support overriding)

**Comment by connorkuehl** [comment]
Since the clusters.provisioning.cattle.io object is now the source of truth for containerd registries, should we:

1. Update consumers to push/pull from clusters.provisioning.cattle.io instead of the containerd-registry setting? (for example, the Harvester UI operates on the containerd-registry setting but it should be updated to operate on the clusters object); OR
2. Have a controller monitor changes to clusters.provisioning.cattle.io containerd registries and update the containerd-registry setting and vice-versa to keep them in sync with each other? This way consumers would not have to be updated to push/pull from clusters.provisioning.cattle.io.

cc: @bk201 @guangbochen 

Aside from that, I am considering implementing this as part of the upgrade scripts to patch the clusters.provisioning.cattle.io object's registry fields with the contents of /etc/rancher/rke2/registries.yaml (if it exists). As far as I can tell, the registries field on that object has existed in Rancher since Rancher v.2.6.0, and Harvester has been using an embedded version of Rancher >= Rancher v2.6.0 since Harvester >= v.0.3.0,  so I think this method should be backwards-compatible with all Harvester installations that support upgrading.

**Comment by bk201** [comment]
Thanks, @connorkuehl.
We only expect users to edit `containerd_registry` setting, so option 2 is much more feasible IMO (come users already rely on this setting). And we probably don't need to sync back from the cluster object to the `containerd_registry` setting.

We can patch the object before `upgrade_rancher`: https://github.com/harvester/harvester/blob/b0530a4dada2247b6367bf34f709ad59156597cc/package/upgrade/upgrade_manifests.sh#L1249

**Comment by connorkuehl** [comment]
@bk201 I don't think this problem is as simple as it first appears.

The structure of the `containerd-registry` setting == the contents of `registries.yaml`, however, that structure is only partially compatible with the structure that the Rancher Cluster's Registry field. The main problem is that the Cluster's Registry field has a different structure for the Configs. There's no problem from what I can see with the Mirrors field.

So if we patched the clusters.provisioning.cattle.io object using the registries.yaml file as input, we will only correctly add the "mirrors" field, and the "configs" field will be lost, since we can't get the correct information that Rancher expects just from the `registries.yaml` file.

For example:

| containerd-registry setting | registries.yaml | clusters.provisioning.cattle.io .spec.rkeConfig.registries |
| - | - | - |
| [wharfie registries.RegistryConfig](https://pkg.go.dev/github.com/rancher/wharfie/pkg/registries#RegistryConfig) | [wharfie registries.RegistryConfig](https://pkg.go.dev/github.com/rancher/wharfie/pkg/registries#RegistryConfig) | [rkev1.RegistryConfig](https://pkg.go.dev/github.com/rancher/rancher/pkg/apis@v0.0.0-20230712180354-5f98b0f28bc8/rke.cattle.io/v1#RegistryConfig) |

It looks like Rancher can derive a valid `registries.yaml` using its Cluster's Registry field, but we cannot derive a valid Registries struct for the Cluster object from `registries.yaml`/`containerd-registry`, since we need to be able to map the inputs from `containerd-registry` setting to the Kubernetes secrets that the rkev1.RegistryConfig needs to refer to.

For example, you can kind of see in the patch where things go off the rails (ctrl+f "TODO"): https://github.com/connorkuehl/harvester/commit/4c0cc82c5319eb451fdb5ed71f7a35899f2b986e

---

Is there a link to where this conversation took place:

> We would expect you to define the registries through your [clusters.provisioning.cattle.io](http://clusters.provisioning.cattle.io/) object and let Rancher designate them -- otherwise, you will need to override the registry with machineGlobalConfig (if possible, it may not actually be possible but I need to look at what we support overriding)

Perhaps I can dive into the machineGlobalConfig approach if the current approach is dead on arrival?

**Comment by connorkuehl** [comment]
Just updating here: I double-checked the machineGlobalConfig and confirmed through experimentation that road is not open to us (this was suggested in the conversation thread as well but I thought I'd just confirm).

---

I believe I found a way forward with the original approach, but it adds some complexity, and I am going to try to implement and test it today, but depending on how testing or reviews go it might not be merged in time for RC4.

I've confirmed that Harvester _does_ have access to the necessary information that the clusters.provisioning.cattle.io object needs, it's just the Clusters object wants to consume it from a Kubernetes secret. So the Harvester setting controller will have to manage the lifetime of Kubernetes Secret objects and put the user, password, auth, identity token config fields into the secret and then when it updates the Cluster object it just tells the Cluster object which secret to go get those values from.

This also complicates the change to the upgrade script since the Kubernetes Secrets juggling doesn't seem tenable with a `yq` one-liner and `kubectl patch` :thinking: 

cc: @bk201 @guangbochen @markhillgit

**Comment by TachunLin** [comment]
Verified **fixed** on `v1.2-a40c4f2f-head` (08/01). Close this issue.

### Result

After upgrade to v1.2-head and fresh install Harvester cluster, 
The containerd registry file content remain exists and the settings also keeps as expected


1. <details><summary>Upgrade from v1.1.2 to v1.2-head $~~$ $\color{green}{\textsf{PASS}}$</summary>
   
      After upgrade to v1.2-head:
      
      1. The containerd registry file and content exists `/etc/rancher/rke2/registries.yaml`
      ```
      harv112upd:~ # cat /etc/rancher/rke2/registries.yaml
      {"configs":{"myregistry.local:5000":{"auth":{"username":"testuser","password":"testpassword","auth":"","identity_token":""},"tls":null}},"mirrors":{"docker.io":{"endpoint":["https://myregistry.local:5000"]}}}
      ```
      2. The Harvester containerd registry setting exists 
      
        ![image](https://github.com/harvester/harvester/assets/29251855/d1a48d3a-238f-4417-ae2b-b5db1cfd86c0)

      3. The rancher version have been upgrade to v2.7.5
      ```
          harv112upd:~ # kubectl get nodes -o json | grep rancher
    "docker.io/rancher/rancher:v2.7.5"
      ```


   </details>

2. <details><summary>Fresh install v1.2-head cluster $~~$ $\color{green}{\textsf{PASS}}$</summary>

      On the fresh install two nodes Harveste v1.2-head cluster, after we set the containerd registry in settings
      
      1. The containerd registry file and content exists in /etc/rancher/rke2/registries.yaml
      
          ```
          n1-bp12:~ # cat /etc/rancher/rke2/registries.yaml
          {"configs":{"myregistry.local:5000":{"auth":{"username":"testuser","password":"testpassword","auth":"","identity_token":""},"tls":null}},"mirrors":{"docker.io":{"endpoint":["https://myregistry.local:5000"]}}}
          ```
      2. The Harvester containerd registry setting exists 
      

      3. The rancher version have been upgrade to v2.7.5
      ```
      n1-bp12:~ # kubectl get nodes -o json | grep rancher
      "docker.io/rancher/rancher:v2.7.5"
      ```
     4. We can set the additonal-ca for the private registry
     
        ![image](https://github.com/harvester/harvester/assets/29251855/25bb28fb-c217-48d7-bac5-300a637cc797)
        
        ![image](https://github.com/harvester/harvester/assets/29251855/f1539cf9-0e73-4697-8144-6221a72e3af2)
        
        ![image](https://github.com/harvester/harvester/assets/29251855/9e874944-66f2-493e-b63d-1b57d5ebfe38)

     5. We can create the nginx deployment to pull image from the private registry 
     
        ![image](https://github.com/harvester/harvester/assets/29251855/375ad30d-cb01-4684-a049-e62384d64cc1)


   </details>



### Test Information
* Test Environment: Single and two nodes Harvester on local kvm machine
* Harvester version: `v1.2-a40c4f2f-head` (08/01)

### Verify Steps
Refer to test steps in:

- [Manual test case](https://harvester.github.io/tests/manual/_incoming/2176-airgap-private-registry-rancher-agent-image/)
- https://github.com/harvester/harvester/issues/2176#issuecomment-1266877245


1. <details><summary>Upgrade from v1.1.2 to v1.2-head</summary>
      
      1. Prepare a v1.1.2 cluster
      2. Prepare another vm for the private registry
      3. Setup private registry and secrets
          - [Manual test case](https://harvester.github.io/tests/manual/_incoming/2176-airgap-private-registry-rancher-agent-image/)
          - https://github.com/harvester/harvester/issues/2176#issuecomment-1266877245
      4. Add some mirrors in the settings -> containerd-registry
          
          ![image](https://github.com/harvester/harvester/assets/29251855/0ec379fc-842b-4805-a0a7-94362cae46b1)

      5. Check the rancher version
          ```
          harv112upd:~ # kubectl get nodes -o json | grep rancher
          docker.io/rancher/rancher:v2.6.11
          ```
      6. Upgrade to v1.2-head
      7. Check the content in file /etc/rancher/rke2/registries.yaml
      8. Check the containerd-registry  settings
     

</details> 

2. <details><summary>Fresh install v1.2-head cluster</summary>
      
      1. Prepare the two node fersh installed Harvester v1.2-head cluster
      2. Prepare another vm for the private registry
      3. Setup private registry and secrets
          - [Manual test case](https://harvester.github.io/tests/manual/_incoming/2176-airgap-private-registry-rancher-agent-image/)
          - https://github.com/harvester/harvester/issues/2176#issuecomment-1266877245
      4. Copy certs/domain.crt content in myregistry VM and paste it to additional-ca setting
      5. Add some mirrors in the settings -> containerd-registry
          
          ![image](https://github.com/harvester/harvester/assets/29251855/0ec379fc-842b-4805-a0a7-94362cae46b1)

      6. Check the rancher version
          ```
          # kubectl get nodes -o json | grep rancher
          ```
      7. Check the content in file /etc/rancher/rke2/registries.yaml
      8. Check the containerd-registry  settings
      9. Open K9s -> search -> secrets -> containerd -> y
      10. Check content in cattle-system/harvester-containerd-registry secret is changed.
      11. Search jobs -> containerd -> l -> tail
      12. Check there are new jobs to auto apply the new containerd-registry setting.
      13. Apply following yaml in file
      ```
      apiVersion: apps/v1
      kind: Deployment
      metadata:
        name: myregistry-nginx
      spec:
        selector:
          matchLabels:
            image: myregistry-nginx
        template:
          metadata:
            labels:
              image: myregistry-nginx
          spec:
            containers:
              - imagePullPolicy: Always
                image: myregistry.local:5000/nginx:latest
                name: nginx
      ```
      ```
      # kubectl apply -f deploy.yaml
      deployment.apps/myregistry-nginx created
      
      ```
      14. search deployments -> default namespace -> myregsistry-nginx -> describe -> shift + g
      15. Check the nginx can be deployed.
          
      
        
  
</details> 




### Additional Context
After upgrade, we need to reset the additional-ca to trigger the jobs
